dicts/dicts.gobra

```diff
@@ -0,0 +1,605 @@
+/*
+  This file is part of gobra-libs which is released under the MIT license.
+  See LICENSE or go to https://github.com/viperproject/gobra-libs/blob/main/LICENSE
+  for full license details.
+
+  This file is inspired by the standard libraries and axiomatisations of the following verifiers:
+  - dafny-lang/dafny: https://github.com/dafny-lang/dafny/blob/master/Source/DafnyStandardLibraries/src/Std/Collections/Map.dfy
+  - verus-lang/verus:
+    - https://github.com/verus-lang/verus/blob/main/source/pervasive/map_lib.rs
+    - https://github.com/verus-lang/verus/blob/main/source/pervasive/map.rs
+  - viperproject/silicon: https://github.com/viperproject/silicon/blob/master/src/main/resources/dafny_axioms/maps.vpr
+  - why3:
+    - https://www.why3.org/stdlib/fmap.html
+    - https://www.why3.org/stdlib/map.html
+*/
+
+// TODO (at the end) Organize definitions: Remove duplicates, merge similar ones, place related ones together
+
+// QUES I plan on making the naming of variables consistent (e.g. keys and values: x/y, or k/v, or key/value);
+// Is there any convention that is particularly "good"?
+
+// This package defines lemmas for mathematical maps (aka dictionaries) commonly used
+// in specifications.
+package dicts
+
+import "util"
+import "sets"
+
+/// Dafny
+// Skip Get: No Option<>
+// Skip ToImap: No Imap
+
+// A dictionary is empty if its domain is empty.
+ghost
+decreases
+pure func IsEmpty(d dict[int]int) bool {
+	return sets.IsEmpty(domain(d))
+}
+
+// Returns the empty dictionary.
+ghost
+ensures IsEmpty(result)
+decreases
+pure func Empty() (result dict[int]int) {
+	return dict[int]int{}
+}
+
+// QUES We structure Remove, RemoveKeys and Restrict a bit differently
+// than Dafny. Are the differences ok?
```

Comment:
I think the most significant difference is that Remove and Restrict are defined in terms of RemoveKeys, while we define Remove and RemoveKeys in terms of Restrict. This is mainly because I didn't know how (or whether it is possible) to implement RemoveKeys in Gobra.